File: code/rust/ch05/src/binary/reader.rs
use crate::binary::chunk;

pub struct Reader {
    data: Vec<u8>,
    pos: usize,
}

impl Reader {
    pub fn new(data: Vec<u8>) -> Reader {
        Reader { data: data, pos: 0 }
    }

    pub fn read_byte(&mut self) -> u8 {
        let b = self.data[self.pos];
        self.pos += 1;
        b
    }

    fn read_u32(&mut self) -> u32 {
        let a0 = self.read_byte() as u32;
        let a1 = self.read_byte() as u32;
        let a2 = self.read_byte() as u32;
        let a3 = self.read_byte() as u32;
        (a3 << 24) | (a2 << 16) | (a1 << 8) | a0
    }

    fn read_u64(&mut self) -> u64 {
        let a0 = self.read_u32() as u64;
        let a1 = self.read_u32() as u64;
        (a1 << 32) | a0
    }

    fn read_lua_integer(&mut self) -> i64 {
        self.read_u64() as i64
    }

    fn read_lua_number(&mut self) -> f64 {
        use std::f64; // TODO
        f64::from_bits(self.read_u64())
    }

    fn read_bytes(&mut self, n: usize) -> Vec<u8> {
        let mut vec = Vec::new();
        for _i in 0..n {
            vec.push(self.read_byte());
        }
        vec
    }

    fn read_string(&mut self) -> String {
        self.read_string0().unwrap_or_else(|| String::new())
    }

    fn read_string0(&mut self) -> Option<String> {
        let mut size = self.read_byte() as usize;
        if size == 0 {
            return None;
        }
        if size == 0xFF {
            size = self.read_u64() as usize; // size_t
        }
        let bytes = self.read_bytes(size - 1);
        let string = String::from_utf8(bytes);
        string.ok() // Some(string.unwrap())
    }

    fn read_vec<T, F>(&mut self, f: F) -> Vec<T>
    where
        F: Fn(&mut Reader) -> T,
    {
        let n = self.read_u32() as usize;
        let mut vec = Vec::with_capacity(n);
        for _i in 0..n {
            vec.push(f(self));
        }
        vec
    }

    pub fn check_header(&mut self) {
        assert_eq!(self.read_bytes(4), chunk::LUA_SIGNATURE, "not a precompiled chunk!");
        assert_eq!(self.read_byte(), chunk::LUAC_VERSION, "version mismatch!");
        assert_eq!(self.read_byte(), chunk::LUAC_FORMAT, "format mismatch!");
        assert_eq!(self.read_bytes(6), chunk::LUAC_DATA, "corrupted!");
        assert_eq!(self.read_byte(), chunk::CINT_SIZE, "int size mismatch!");
        assert_eq!(self.read_byte(), chunk::CSIZET_SIZE, "size_t size mismatch!");
        assert_eq!(self.read_byte(), chunk::INSTRUCTION_SIZE, "instruction size mismatch!");
        assert_eq!(self.read_byte(), chunk::LUA_INTEGER_SIZE, "lua_Integer size mismatch!");
        assert_eq!(self.read_byte(), chunk::LUA_NUMBER_SIZE, "lua_Number size mismatch!");
        assert_eq!(self.read_lua_integer(), chunk::LUAC_INT, "endianness mismatch!");
        assert_eq!(self.read_lua_number(), chunk::LUAC_NUM, "float format mismatch!");
    }

    pub fn read_proto(&mut self) -> chunk::Prototype {
        self.read_proto0(None)
    }

    fn read_proto0(&mut self, parent_source: Option<String>) -> chunk::Prototype {
        let source = self.read_string0().or(parent_source);
        chunk::Prototype {
            source: source.clone(), // debug
            line_defined: self.read_u32(),
            last_line_defined: self.read_u32(),
            num_params: self.read_byte(),
            is_vararg: self.read_byte(),
            max_stack_size: self.read_byte(),
            code: self.read_vec(|r| r.read_u32()),
            constants: self.read_vec(|r| r.read_constant()),
            upvalues: self.read_vec(|r| r.read_upvalue()),
            protos: self.read_vec(|r| r.read_proto0(source.clone())),
            line_info: self.read_vec(|r| r.read_u32()),        // debug
            loc_vars: self.read_vec(|r| r.read_loc_var()),     // debug
            upvalue_names: self.read_vec(|r| r.read_string()), // debug
        }
    }

    fn read_constant(&mut self) -> chunk::Constant {
        let tag = self.read_byte();
        match tag {
            chunk::TAG_NIL => chunk::Constant::Nil,
            chunk::TAG_BOOLEAN => chunk::Constant::Boolean(self.read_byte() != 0),
            chunk::TAG_INTEGER => chunk::Constant::Integer(self.read_lua_integer()),
            chunk::TAG_NUMBER => chunk::Constant::Number(self.read_lua_number()),
            chunk::TAG_SHORT_STR => chunk::Constant::Str(self.read_string()),
            chunk::TAG_LONG_STR => chunk::Constant::Str(self.read_string()),
            _ => panic!("corrupted!"),
        }
    }

    fn read_upvalue(&mut self) -> chunk::Upvalue {
        chunk::Upvalue {
            instack: self.read_byte(),
            idx: self.read_byte(),
        }
    }

    fn read_loc_var(&mut self) -> chunk::LocVar {
        chunk::LocVar {
            var_name: self.read_string(),
            start_pc: self.read_u32(),
            end_pc: self.read_u32(),
        }
    }
}

File: code/csharp/ch09/luavm/binchunk/Reader.cs
using System;
using System.Linq;
using luavm.util;

namespace luavm.binchunk
{
    internal class Reader
    {
        public byte[] data;

        public byte readByte()
        {
            var b = data[0];
            data = data.Skip(1).ToArray();
            return b;
        }

        uint readUint32()
        {
            var bytes = new byte[4];
            Array.ConstrainedCopy(data, 0, bytes, 0, 4);
            data = data.Skip(4).ToArray();
            if (!BitConverter.IsLittleEndian)
            {
                Array.Reverse(bytes);
            }

            return BitConverter.ToUInt32(bytes, 0);
        }

        long readLuaInteger()
        {
            return (long) readUint64();
        }

        double readLuaNumber()
        {
            return BitConverter.Int64BitsToDouble((long) readUint64());
        }

        private ulong readUint64()
        {
            var bytes = new byte[8];
            Array.ConstrainedCopy(data, 0, bytes, 0, 8);
            data = data.Skip(8).ToArray();
            if (!BitConverter.IsLittleEndian)
            {
                Array.Reverse(bytes);
            }

            return BitConverter.ToUInt64(bytes, 0);
        }

        private byte[] readBytes(uint n)
        {
            var bytes = new byte[n];
            Array.ConstrainedCopy(data, 0, bytes, 0, (int) n);
            data = data.Skip((int) n).ToArray();
            return bytes;
        }

        public void checkHeader()
        {
//            Console.WriteLine(Encoding.Default.GetString(readBytes(4)));
            if (ConvertUtil.Bytes2String(readBytes(4)) != BinaryChunk.LUA_SIGNATURE)
            {
                Console.WriteLine("not a precompiled chunk!");
            }

            if (readByte() != BinaryChunk.LUAC_VERSION)
            {
                Console.WriteLine("version mismatch!");
            }

            if (readByte() != BinaryChunk.LUAC_FORMAT)
            {
                Console.WriteLine("format mismatch!");
            }

            if (ConvertUtil.Bytes2String(readBytes(6)) != BinaryChunk.LUAC_DATA)
            {
                Console.WriteLine("corrupted!");
            }

            if (readByte() != BinaryChunk.CINT_SIZE)
            {
                Console.WriteLine("int size mismatch!");
            }

            var b = readByte();
            if (b != BinaryChunk.CSIZET_SIZE_32 && b != BinaryChunk.CSIZET_SIZE_64)
            {
                Console.WriteLine("size_t size mismatch!");
            }

            if (readByte() != BinaryChunk.INSTRUCTION_SIZE)
            {
                Console.WriteLine("instruction size mismatch!");
            }

            if (readByte() != BinaryChunk.LUA_INTEGER_SIZE)
            {
                Console.WriteLine("lua_Integer size mismatch!");
            }

            if (readByte() != BinaryChunk.LUA_NUMBER_SIZE)
            {
                Console.WriteLine("lua_Number size mismatch!");
            }

            if (readLuaInteger() != BinaryChunk.LUAC_INT)
            {
                Console.WriteLine("endianness mismatch!");
            }

            if (!readLuaNumber().Equals(BinaryChunk.LUAC_NUM))
            {
                Console.WriteLine("float format mismatch!");
            }
        }

        public Prototype readProto(string parentSource)
        {
            var source = readString();
            if (source == "")
            {
                source = parentSource;
            }

            return new Prototype
            {
                Source = source,
                LineDefined = readUint32(),
                LastLineDefined = readUint32(),
                NumParams = readByte(),
                IsVararg = readByte(),
                MaxStackSize = readByte(),
                Code = readCode(),
                Constants = readConstants(),
                Upvalues = readUpvalues(),
                Protos = readProtos(source),
                LineInfo = readLineInfo(),
                LocVars = readLocVars(),
                UpvalueNames = readUpvalueNames()
            };
        }

        private string[] readUpvalueNames()
        {
            var names = new string[readUint32()];
            for (var i = 0; i < names.Length; i++)
            {
                names[i] = readString();
            }

            return names;
        }

        private LocVar[] readLocVars()
        {
            var locVars = new LocVar[readUint32()];
            for (var i = 0; i < locVars.Length; i++)
            {
                locVars[i] = new LocVar
                {
                    VarName = readString(),
                    StartPC = readUint32(),
                    EndPC = readUint32()
                };
            }

            return locVars;
        }

        private uint[] readLineInfo()
        {
            var lineInfo = new uint[readUint32()];
            for (var i = 0; i < lineInfo.Length; i++)
            {
                lineInfo[i] = readUint32();
            }

            return lineInfo;
        }

        private Prototype[] readProtos(string parentSource)
        {
            var protos = new Prototype[readUint32()];
            for (var i = 0; i < protos.Length; i++)
            {
                protos[i] = readProto(parentSource);
            }

            return protos;
        }

        private Upvalue[] readUpvalues()
        {
            var upvalues = new Upvalue[readUint32()];
            for (var i = 0; i < upvalues.Length; i++)
            {
                upvalues[i] = new Upvalue
                {
                    Instack = readByte(),
                    Idx = readByte()
                };
            }

            return upvalues;
        }

        private object[] readConstants()
        {
            var constants = new object[readUint32()];
            for (var i = 0; i < constants.Length; i++)
            {
                constants[i] = readConstant();
            }

            return constants;
        }

        private object readConstant()
        {
            switch (readByte())
            {
                case BinaryChunk.TAG_NIL: return null;
                case BinaryChunk.TAG_BOOLEAN: return readByte() != 0;
                case BinaryChunk.TAG_INTEGER: return readLuaInteger();
                case BinaryChunk.TAG_NUMBER: return readLuaNumber();
                case BinaryChunk.TAG_SHORT_STR: return readString();
                case BinaryChunk.TAG_LONG_STR: return readString();
                default: throw new Exception("corrupted!");
            }
        }

        private uint[] readCode()
        {
            var code = new uint[readUint32()];
            for (var i = 0; i < code.Length; i++)
            {
                code[i] = readUint32();
            }

            return code;
        }

        private string readString()
        {
            var size = (uint) readByte();
            if (size == 0)
            {
                return "";
            }

            if (size == 0xFF)
            {
                size = (uint) readUint64();
            }

            var bytes = readBytes(size - 1);
            return ConvertUtil.Bytes2String(bytes);
        }
    }
}